### PR TITLE
Initial Testable Implementation

### PR DESCRIPTION
**Pull request by jessebmiller**

### CODE REVIEW

File: auth/auth.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package auth
+
+type Auth struct {
+	Permissionser Permissionser
+	ConstraintRunner ConstraintRunner
+}
+
+func isIn(a string, xs []string) bool {
```
Comment (henryrizzi) [inline auth/auth.go]: know this is a generic, but the variable names this subset check a bit unclear.

File: auth/auth.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	for _, constraint := range constraints {
+		if (!cr.Run(constraint, ctx)) {
+			return false
+		}
+	}
+	return true
+}
+
+// May is true iff we can find any relevant permission whose constraints pass
+func (a Auth) May(c Context) bool {
```
Comment (henryrizzi) [inline auth/auth.go]: a->auth, c->ctx

File: auth/context.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package auth
+
+type Context interface {
+	Action() string
+	ResourceType() string
+	Roles() []string
+}
+
```
Comment (henryrizzi) [inline auth/context.go:8]: Separate out literal context into concrete implementation file possibly

File: auth/providers.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
+package auth
+
```
Comment (henryrizzi) [inline auth/providers.go:2]: document: these are meant to be metadata driven.

File: auth/providers.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+type Permission struct {
+	Actions       []string
+	ResourceTypes []string
+	Constraints   []string
+}
+
+type Permissionser interface {
+	Permissions([]string) []Permission // Roles to Permissions
```
Comment (henryrizzi) [inline auth/providers.go]: TODO applies here as well from below.

File: auth/providers.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	// TODO this might error, handle that..
+	Run(string, Context) bool
+}
+
+/*
+ * builtin implementations
+ */
+
+type funcMapConstraintRunner struct {
+	constraintFuncs map[string]func(Context) bool
```
Comment (henryrizzi) [inline auth/providers.go]: Context can have anything, therefore context can have enough information to validate.
Document: Client code is responsible for having context and constraints that agree with each other.

File: auth/providers.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+func (fmcr funcMapConstraintRunner) Run (constraint string, ctx Context) bool {
+	return fmcr.constraintFuncs[constraint](ctx)
+}
+
+func NewFuncMapConstraintRunner(fm map[string]func(Context) bool) funcMapConstraintRunner {
```
Comment (henryrizzi) [inline auth/providers.go]: java: new Object, go: NewObject

File: auth/providers.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	// role: permissions granted to that role
+	permissions map[string][]Permission
+
+}
+
+func (lp mapPrivileges) Permissions(roles []string) []Permission {
+	// gather the permissions granted to each given role
+	perms := []Permission{}
+	for _, r := range roles {
+		perms = append(perms, lp.permissions[r]...)
```
Comment (henryrizzi) [inline auth/providers.go]: ... = *

File: auth/providers.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+	return funcMapConstraintRunner{fm}
+}
+
+type mapPrivileges struct {
+	// role: permissions granted to that role
+	permissions map[string][]Permission
+
+}
+
+func (lp mapPrivileges) Permissions(roles []string) []Permission {
```
Comment (henryrizzi) [inline auth/providers.go]: mp

File: auth/auth_test.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+			[]string{} },
+	}
+	testPermissionser := mockPermissionser{ "tester", testPermissions }
+	unconstrainedRunner := constantConstraintRunner{ true }
+	constrainedRunner := constantConstraintRunner{ false }
+	unconstrainedAuth := auth.Auth{ testPermissionser, unconstrainedRunner }
+	constrainedAuth := auth.Auth{ testPermissionser, constrainedRunner }
+	authorizedContext := mockContext{ "A0", "R1", []string{"tester"} }
+	unauthorizedContext := mockContext{ "AX", "R0", []string{"tester"} }
+	wrongRoleContext := mockContext{ "A0", "R1", []string{"not-tester"} }
```
Comment (henryrizzi) [inline auth/auth_test.go]: "not-tester" won't match the role which the mock permissionser expects.

File: main.go [hunk 11]
```diff
@@ -1,7 +1,23 @@
...
+)
+
+type ProxyConfig struct {
+	ListenPort int
+	backendURL url.URL
+}
 
 func main() {
-	fmt.Println("Running trbac")
+	log.Println("Trbac proxy starting up...")
```
Comment (henryrizzi) [inline main.go]: the stuff in here could be removed, but nbd

**Review by henryrizzi — COMMENTED** [review]
Just a few comments, more requests for documentation and clearer naming convention than things that actually need to be changed before a merge.

File: auth/providers.go [hunk 12]
```diff
@@ -0,0 +1,75 @@
+package auth
+
+/*
+
+Providers are inteded to be metadat driven.
```
Comment (henryrizzi) [inline auth/providers.go:5]: metadata

File: auth/auth_test.go [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+		auth.Permission{
+			[]string{"A0", "A1"},     // Actions
+			[]string{"R0", "R1"},     // ResourceTypes
+			[]string{"C0", "C1"} },   // Constraints
+		auth.Permission{
+			[]string{"AX"},
+			[]string{"RX", "RY"},
+			[]string{} },
+	}
+	testPrivileges := mockPrivileges{ "tester", testPermissions }
```
Comment (henryrizzi) [inline auth/auth_test.go]: Legibility notes: newline after

File: auth/auth_test.go [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+	unconstrainedRunner := constantConstraintRunner{ true }
+	constrainedRunner := constantConstraintRunner{ false }
```
Comment (henryrizzi) [inline auth/auth_test.go]: newline after

File: auth/auth_test.go [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+	unconstrainedAuth := auth.Auth{ testPrivileges, unconstrainedRunner }
+	constrainedAuth := auth.Auth{ testPrivileges, constrainedRunner }
```
Comment (henryrizzi) [inline auth/auth_test.go]: newline after

File: auth/auth_test.go [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+	authorizedContext := mockContext{ "A0", "R1", []string{"tester"} }
+	unauthorizedContext := mockContext{ "AX", "R0", []string{"tester"} }
+	wrongRoleContext := mockContext{ "A0", "R1", []string{"not-tester"} }
```
Comment (henryrizzi) [inline auth/auth_test.go]: newline after

**Review by henryrizzi — COMMENTED** [review]
Just formatting suggestions and one spelling thing

File: auth/providers.go (hunk 12 above)
Comment (jessebmiller) [reply auth/providers.go:5]: METADAT is my kraftwork cover band

File: auth/auth_test.go (hunk 13 above)
Comment (jessebmiller) [reply auth/auth_test.go]: I should actually just run go fmt

File: auth/auth_test.go (hunk 14 above)
Comment (jessebmiller) [reply auth/auth_test.go]: again... go fmt

File: auth/providers.go (hunk 12 above)
Comment (henryrizzi) [reply auth/providers.go:5]: sehr gut!

**Pull request merged**